crates/red_knot_python_semantic/src/types/infer.rs

```diff
@@ -1038,28 +1042,74 @@ impl<'db> TypeInferenceBuilder<'db> {
         parameter_with_default: &ast::ParameterWithDefault,
         definition: Definition<'db>,
     ) {
-        // TODO(dhruvmanila): Infer types from annotation or default expression
-        // TODO check that default is assignable to parameter type
-        self.infer_parameter_definition(&parameter_with_default.parameter, definition);
+        let ast::ParameterWithDefault {
+            parameter,
+            default,
+            range: _,
+        } = parameter_with_default;
+        let opt_default_ty = default
+            .as_ref()
+            .map(|default| definition_expression_ty(self.db, definition, default));
```

Comment:
Just an understand question. Why are we calling `definition_expression_ty` over `infer_standalone_expression`? 

Do we need to merge the diagnostics from the `definition_expression_ty` ?

Reply:
> Why are we calling `definition_expression_ty` over `infer_standalone_expression`?

Because parameter default values are not marked as standalone expressions in the semantic index builder. We minimize the expressions that are marked as standalone expressions, to reduce ingredient counts. In this case we don't need them to be standalone expressions, we can use `definition_expression_ty` to find the right scope, infer that scope, and pull the expression type out.

> Do we need to merge the diagnostics from the `definition_expression_ty` ?

No, in fact we shouldn't. Function parameter definitions are a little odd, in that the definition is part of the function body scope (it defines a symbol in the function body), but the expressions whose type we need here (the annotation and the default) are not part of the function body scope, they are part of the outer scope (or possibly two different outer scopes: the defaults are always in the scope the function is defined in, the annotations might be in the intervening type-parameters scope, for a generic function.) Part of the functionality of `definition_expression_ty` is that it can handle inferring expressions that are in some other scope. But we shouldn't try to merge diagnostics that belong to a different scope.

Reply:
I realized here that `definition_expression_ty` is kind of over-powered for this use case, since in these cases we know for sure that the expression is not in the same scope as the definition. So I've instead introduced an `expression_ty` that just looks up the right scope for the expression, infers that scope, and pulls out the expression type.

Reply:
Is there a use-case for having `expression_ty` as a standalone function instead of adding it as a method on `TypeInference`? For the latter, we could add an assert to make sure that the expression is not from the same scope as the one that's currently being inferred which is available via the `self.scope` method:

https://github.com/astral-sh/ruff/blob/56a631a86891167ad1af8c10ae3751976bbeca6d/crates/red_knot_python_semantic/src/types/infer.rs#L393-L395

Reply:
That's a good idea, I'll put up a PR for this.

Reply:
https://github.com/astral-sh/ruff/pull/14879